### PR TITLE
92 - conditional billing address

### PR DESCRIPTION
**Pull request by jachan**
## Link to issue
https://github.com/newjersey/doula-pm/issues/92

## What was done?
- Add question to Personal information 2 asking if billing and mailing address are the same
- Add Billing Address fields on Personal Information 2
- Add logic on download page to overwrite billing address fields if billing and mailing address are the same
- Add logic to PDF generation to fill in page 5 and 7 fields with pay to address
- Unit tests
- Change copy from `Street Address 1` to `Street Address` to match designs
- Add "Billing" to error messages (e.g. Billing zip is required), as mocks do not specify and it would be confusing for a non-sighted user

## Accessibility

- [x] I have made sure this works with a screenreader!
- [x] I have checked this with the [WAVE extension](https://wave.webaim.org/extension/).

## How should a reviewer test?

- Clone down repo, `npm run test`
- `npm run dev`
- Verify that show/hide logic for yes/no works correctly
- Verify that moving to next screen and returning keeps the radio button filled in properly
- Verify that moving to the end of the flow downloads the PDF with Pay To address filled in on Pages 5 and 7
- Verify that answering "No", filling in address and then changing answer to "Yes" populates PDF with same Mailing/Pay To address

## Screenshots (for visual changes)

- Before
image

- After
image
image
image
image

### CODE REVIEW

File: src/app/form/(formSteps)/download/page.tsx [hunk 1]
```diff
@@ -46,6 +48,18 @@ const getFormData = (): FormData => {
     city: getValue("city"),
     state: state,
     zip: getValue("zip"),
+    hasSameBillingMailingAddress: hasSameBillingMailingAddress,
```
Comment (jachan) [inline src/app/form/(formSteps)/download/page.tsx:51]: Unclear to me whether this logic should live here or in the PDF generation section.

File: src/app/form/(formSteps)/download/page.tsx [hunk 2]
```diff
@@ -56,7 +70,7 @@ const getFormData = (): FormData => {
   };
 };
 
-const DownloadStep = () => {
+const DownloadSection = () => {
```
Comment (jachan) [inline src/app/form/(formSteps)/download/page.tsx:73]: Renamed to be more consistent with the current way we're using these terms

File: src/app/form/(formSteps)/personal-details/2/PersonalDetailsStep2.test.tsx [hunk 3]
```diff
@@ -8,20 +8,50 @@ import { render, screen, within } from "@testing-library/react";
 import userEvent from "@testing-library/user-event";
 import { type AppRouterInstance } from "next/dist/shared/lib/app-router-context.shared-runtime";
 
+const clickNoSameBillingMailingAddress = async () => {
```
Comment (mluedke2) [inline src/app/form/(formSteps)/personal-details/2/PersonalDetailsStep2.test.tsx:11]: [dust] instead of `clickNoSameBillingMailingAddress` and `clickYesSameBillingMailingAddress`, maybe `clickSameBillingMailingAddressNo` and `clickSameBillingMailingAddressYes`? So only their suffixes are different?

File: src/app/form/(formSteps)/download/DownloadSection.test.tsx [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        await waitFor(() => {
+          expect(mockFillAllForms).toHaveBeenCalledWith(
+            expect.objectContaining({
+              dateOfBirth: new Date("12/25/1990"),
+            }),
+          );
+        });
+      });
+
+      it("sets dateOfBirth to null when any date component is missing", async () => {
```
[the remark this answers is not in the document]
Comment (jachan) [reply src/app/form/(formSteps)/download/DownloadSection.test.tsx:161]: This is interesting, because if our validation is working properly, date should never be empty! This makes me want to throw an error since it's an unexpected state.

However, for the user, the error is functionally useless, and may be actively unhelpful if it stops them from completing the application. I think it might be more helpful from a user perspective to skip any fields with errors and fill in the rest of the PDF. What do you think?

File: src/app/form/(formSteps)/finish/FinishSection.test.tsx [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        city: "New York",
+        state: "NY",
+        zip: "10001",
+        hasSameBillingMailingAddress: "true",
+        // These billing values should be ignored
+        billingStreetAddress1: "456 Oak Ave",
+        billingStreetAddress2: "Suite 200",
+        billingCity: "Los Angeles",
+        billingState: "CA",
+        billingZip: "90210",
```
Comment (mluedke2) [inline src/app/form/(formSteps)/finish/FinishSection.test.tsx:79]: [dust] - doesn't really affect the test much, but if `hasSameBillingMailingAddress: "true"` then in practice would `billingStreetAddress1` and all the others be `undefined`?

Or is there an intended edge case where the user says "no they're not the same" then fills out a separate billing, and then goes back and says "actually yeah they're the same" and values are left in the fields?

File: src/app/form/(formSteps)/personal-details/2/PersonalDetailsStep2.test.tsx [hunk 6]
```diff
@@ -130,7 +184,7 @@ describe("<PersonalDetailsStep2 />", () => {
       ).toBeInTheDocument();
 
       const expectedErrorMessages = [
-        "Street address 1 is required",
+        "Street address is required",
         "City is required",
         "ZIP code is required",
       ];
```
Comment (josephine-njooi) [inline src/app/form/(formSteps)/personal-details/2/PersonalDetailsStep2.test.tsx:190]: Dust: I had a thought from our error summary design convo: there's a test below for `"clicking on the $name error focuses on the input"`. Should this also be tested for the new fields?

File: src/app/form/(formSteps)/finish/page.tsx [hunk 7]
```diff
@@ -31,6 +31,8 @@ const getFormData = (): FormData => {
     convertToBoolean(getValue("isSoleProprietorship")) === true
       ? DisclosingEntity.SoleProprietorship
       : null;
+  const hasSameBillingMailingAddress =
+    convertToBoolean(getValue("hasSameBillingMailingAddress")) === true ? true : false;
```
[the remark this answers is not in the document]
Comment (jachan) [reply src/app/form/(formSteps)/finish/page.tsx:35]: ConvertToBoolean can return null if given null. Validation should stop the user from submitting a null value here

We have an ESLint rule that requires us to handle nullables explicitly. Proposing that we use the `??` nullish coalescing operator to handle the edge case.

File: src/app/form/(formSteps)/personal-details/2/page.tsx [hunk 8]
```diff
@@ -37,12 +52,20 @@ const PersonalDetailsStep2 = () => {
       city: getValue("city") || "",
       state: getValue("state") || "NJ",
       zip: getValue("zip") || "",
+      hasSameBillingMailingAddress: getValue("hasSameBillingMailingAddress") || "",
```
[the remark this answers is not in the document]
Comment (jachan) [reply src/app/form/(formSteps)/personal-details/2/page.tsx:55]: After discussion today at Eng Sync, I believe we aligned on type-conversion logic as part of `GetFormData`.

File: src/app/form/(formSteps)/personal-details/2/page.tsx [hunk 9]
```diff
@@ -190,6 +213,185 @@ const PersonalDetailsStep2 = () => {
...
+                  <div>
+                    <p className="font-ui-xs text-normal">
+                      Are your billing and residential addresses the same?
+                    </p>
+                    <p className="font-ui-xs text-normal">
+                      Select one <RequiredMarker />
+                    </p>
+                  </div>
+                }
+                legendStyle="large"
```
[the remark this answers is not in the document]
Comment (jachan) [reply src/app/form/(formSteps)/personal-details/2/page.tsx:232]: Good catch, thanks!

File: src/app/form/(formSteps)/personal-details/PersonalDetailsData.ts [hunk 10]
```diff
@@ -16,6 +16,12 @@ export interface PersonalDetails2Data {
   city: string | null;
   state: string | null;
   zip: string | null;
+  hasSameBillingMailingAddress: string | null;
```
[the remark this answers is not in the document]
Comment (jachan) [reply src/app/form/(formSteps)/personal-details/PersonalDetailsData.ts:19]: SessionStorage can only store `string`s. Agreed with the rest of the Doula Engineers that conversion from `string` to `boolean` will happen as part of GetFormData in the `finish` section.

File: src/app/form/(formSteps)/personal-details/2/PersonalDetailsStep2.test.tsx [hunk 11]
```diff
@@ -8,20 +8,50 @@ import { render, screen, within } from "@testing-library/react";
...
 ];
 
+const billingAddressFields = [
+  { name: "Street address *", key: "billingStreetAddress1", testValue: "Test address 1" },
+  { name: "Street address line 2", key: "billingStreetAddress2", testValue: "Test address 2" },
+  { name: "City *", key: "billingCity", testValue: "Test city" },
+  { name: "ZIP code *", key: "billingZip", testValue: "12345" },
+];
+
 const allInputFields: Array<InputField> = [
```
[the remark this answers is not in the document]
Comment (jachan) [reply src/app/form/(formSteps)/personal-details/2/PersonalDetailsStep2.test.tsx:47]: Renamed these fields, curious to see what your thoughts are once the new revision is out for review.

File: src/app/form/(formSteps)/personal-details/2/page.tsx [hunk 12]
```diff
@@ -190,6 +213,185 @@ const PersonalDetailsStep2 = () => {
...
+                aria-invalid={errors.hasSameBillingMailingAddress ? "true" : "false"}
+                aria-describedby={
+                  errors.hasSameBillingMailingAddress && "hasSameBillingMailingAddressErrorMessage"
+                }
+              >
+                {errors.hasSameBillingMailingAddress && (
```
[the remark this answers is not in the document]
Comment (jachan) [reply src/app/form/(formSteps)/personal-details/2/page.tsx:238]: Thanks, fixing the comment above brought this in line with the rest of the inputs.

File: src/app/form/(formSteps)/personal-details/2/page.tsx [hunk 13]
```diff
@@ -190,6 +213,185 @@ const PersonalDetailsStep2 = () => {
...
+                    <p className="font-ui-xs text-normal">
+                      Select one <RequiredMarker />
+                    </p>
+                  </div>
+                }
+                legendStyle="large"
+                aria-invalid={errors.hasSameBillingMailingAddress ? "true" : "false"}
+                aria-describedby={
+                  errors.hasSameBillingMailingAddress && "hasSameBillingMailingAddressErrorMessage"
+                }
```
[the remark this answers is not in the document]
Comment (jachan) [reply src/app/form/(formSteps)/personal-details/2/page.tsx:236]: I think you're right! It should be on the radio buttons. Fixed.

File: src/app/form/(formSteps)/download/DownloadSection.test.tsx [hunk 14]
```diff
@@ -0,0 +1,214 @@
+import * as fillPdfForm from "@form/_utils/fillPdf/form";
+import * as zipForms from "@form/_utils/fillPdf/zip";
+import { AddressState, DisclosingEntity } from "@form/_utils/inputFields/enums";
+import * as sessionStorage from "@form/_utils/sessionStorage";
+import { jest } from "@jest/globals";
+import { render, screen, waitFor } from "@testing-library/react";
+import DownloadSection from "./page";
+
+// Mock the dependencies
```
[the remark this answers is not in the document]
Comment (jachan) [reply src/app/form/(formSteps)/download/DownloadSection.test.tsx]: Agree that these comments should be removed - this test file was me playing around with using Copilot...hence the messy mocks, etc. Appreciate your PR comments.

File: src/app/form/(formSteps)/finish/FinishSection.test.tsx (hunk 5 above)
Comment (jachan) [reply src/app/form/(formSteps)/finish/FinishSection.test.tsx]: It's an intended edge case, as we still write the values to `sessionStorage` when a user shows and then hides the fields.

File: src/app/form/(formSteps)/download/DownloadSection.test.tsx [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+  return function MockFormProgressButtons() {
+    return <div data-testid="form-progress-buttons">Form Progress Buttons</div>;
+  };
+});
+
+const mockGetValue = sessionStorage.getValue as jest.MockedFunction<typeof sessionStorage.getValue>;
+const mockFillAllForms = fillPdfForm.fillAllForms as jest.MockedFunction<
+  typeof fillPdfForm.fillAllForms
+>;
+const mockZipForms = zipForms.zipForms as jest.MockedFunction<typeof zipForms.zipForms>;
```
[the remark this answers is not in the document]
Comment (jachan) [reply src/app/form/(formSteps)/download/DownloadSection.test.tsx]: Took a stab at simplifying in the latest revision, lmk what you think.

**Comment by jachan** [comment]
Seeing this error in the cypress tests 😵‍💫 

>  src/app/form/(formSteps)/finish/page.tsx
Type error: Page "src/app/form/(formSteps)/finish/page.tsx" does not match the required types of a Next.js Page.
  "getFormData" is not a valid Page export field.
  
I think I'm exporting the function for test wrong? Any advice?

File: src/app/form/(formSteps)/personal-details/2/page.tsx [hunk 16]
```diff
@@ -184,6 +207,180 @@ const PersonalDetailsStep2 = () => {
...
+                checked={hasSameBillingMailingAddress === "false"}
+                required
+                {...register("hasSameBillingMailingAddress", {
+                  required: "This question is required",
+                })}
+              />
+            </Fieldset>
+
+            {hasSameBillingMailingAddress === "false" && (
+              <Fieldset
```
[the remark this answers is not in the document]
Comment (josephine-njooi) [reply src/app/form/(formSteps)/personal-details/2/page.tsx]: I think it's sufficient, because this implementation is the example in both USWDS and the trussworks library. But #accessibility-all it would have deeper knowledge than me!

https://trussworks.github.io/react-uswds/?path=/docs/page-templates-forms--docs

File: src/app/form/(formSteps)/finish/FinishSection.test.tsx [hunk 17]
```diff
@@ -0,0 +1,183 @@
+import FinishSection from "@form/(formSteps)/finish/page";
+import { AddressState, DisclosingEntity } from "@form/_utils/inputFields/enums";
+import { RouterPathnameProvider } from "@form/_utils/testUtils";
+import { jest } from "@jest/globals";
+import { render, screen, waitFor } from "@testing-library/react";
+import { type AppRouterInstance } from "next/dist/shared/lib/app-router-context.shared-runtime";
+import { getFormData } from "./getFormData";
+
+jest.mock("@form/_utils/fillPdf/form");
+jest.mock("@form/_utils/fillPdf/zip");
```
Comment (josephine-njooi) [inline src/app/form/(formSteps)/finish/FinishSection.test.tsx:10]: Pebble: I'm still wondering, why do these need to be mocked? it seems that the logic in both files should be fully in-memory js and accessible to the test. Is jest unable to import something?

File: src/app/form/(formSteps)/personal-details/2/page.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (josephine-njooi) [reply src/app/form/(formSteps)/personal-details/2/page.tsx:55]: Ah - I was talking about the session storage -> form data in the finish page to put into a pdf, that that would handle that type conversion. Because everything in session storage needs to be a string

For this comment, I'm talking about whether we think the React form `state` in this component. React states don't need to be strings and I'm wondering if this should be typed as a boolean, same with what Matt mentioned here 🙂 https://github.com/newjersey/doula-medicaid/pull/61#discussion_r2270393300

**Review by josephine-njooi — COMMENTED** [review]
> EDIT: Fixed! Moved getFormData out to separate file.

Ah! Also see https://github.com/newjersey/doula-medicaid/pull/27 🙂

File: src/app/form/(formSteps)/download/DownloadSection.test.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (josephine-njooi) [reply src/app/form/(formSteps)/download/DownloadSection.test.tsx]: Ah! I see 😛.
One idea, is that I know that the Github UI puts me more in the mindset of reviewing changed code that I didn't write, and needing to understand and justify every line. My IDE tends to put me more in a mindset that the written code already exists and and is tech debt, that the default is either to work around it or to have to justify changing it. The stuff that I'm adding/changing in my IDE I justify as I change it, but I'm not usually trying to justify lines that are already in my IDE.
I don't know if this would help at all, but one idea might be to push a branch and run compare/draft PR, and review the copilot code like it was written by someone else (which it is), and like you have no idea what it does. Maybe even leave comments for it 🤷‍♀️

File: src/app/form/(formSteps)/finish/FinishSection.test.tsx (hunk 17 above)
Comment (jachan) [reply src/app/form/(formSteps)/finish/FinishSection.test.tsx:10]: For `form`, 
We have a dependency on `fetch` here that is complex to mock out:
https://github.com/newjersey/doula-medicaid/blob/main/src/app/form/_utils/fillPdf/form.ts#L48 
...and then if we mock out `form`, then `pdfDataList` is undefined for https://github.com/newjersey/doula-medicaid/blob/main/src/app/form/_utils/fillPdf/zip.ts#L11C1-L13C6.

Neither of these are relevant to the logic that's being tested here so I decided to leave them mocked.

File: src/app/form/(formSteps)/finish/FinishSection.test.tsx [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+
+describe("<FinishSection />", () => {
+  beforeEach(() => {
+    window.sessionStorage.clear();
+    window.sessionStorage.setItem("dateOfBirthDay", "25");
+    window.sessionStorage.setItem("dateOfBirthMonth", "12");
+    window.sessionStorage.setItem("dateOfBirthYear", "1990");
+    window.sessionStorage.setItem("hasSameBillingMailingAddress", "true");
+    jest.clearAllMocks();
+    mockCreateObjectURL.mockReturnValue("mock-blob-url");
```
[the remark this answers is not in the document]
Comment (jachan) [reply src/app/form/(formSteps)/finish/FinishSection.test.tsx:38]: Ah, apologies for overlooking this comment, there have been a lot in this PR 😅 Yes, I'm happy to pull the setup out to a separate helper function.

File: src/app/form/(formSteps)/finish/FinishSection.test.tsx (hunk 17 above)
Comment (josephine-njooi) [reply src/app/form/(formSteps)/finish/FinishSection.test.tsx:10]: Got it, thanks!

File: src/app/form/(formSteps)/finish/FinishSection.test.tsx [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+    });
+  });
+
+  describe("date of birth handling", () => {
+    it("creates date when all date components are present", async () => {
+      window.sessionStorage.setItem("dateOfBirthDay", "25");
+      window.sessionStorage.setItem("dateOfBirthMonth", "12");
+      window.sessionStorage.setItem("dateOfBirthYear", "1990");
+
+      renderWithRouter();
```
Comment (josephine-njooi) [inline src/app/form/(formSteps)/finish/FinishSection.test.tsx:129]: Sand: I ran all tests in this file, and happened to see that there was this `console.error`, which was coming from this test and can be reproduced by running this test in isolation: `npm test -- --runTestsByPath "src/app/form/(formSteps)/finish/FinishSection.test.tsx" -t "creates date when all date components are present"`

```sh

  console.error
    An update to FinishSection inside a test was not wrapped in act(...).

    When testing, code that causes React state updates should be wrapped into act(...):

    act(() => {
      /* fire events that update state */
    });
    /* assert on the output */

    This ensures that you're testing the behavior the user would see in the browser. Learn more at https://react.dev/link/wrap-tests-with-act

      13 |       const filledForms = await fillAllForms(getFormData());
      14 |       const zipBlob = await zipForms(filledForms);
    > 15 |       setZipDownloadUrl(URL.createObjectURL(zipBlob));
         |       ^
      16 |     })();
      17 |   }, []);
      18 |

      at node_modules/react-dom/cjs/react-dom-client.development.js:15751:19
      at runWithFiberInDEV (node_modules/react-dom/cjs/react-dom-client.development.js:543:16)
      at warnIfUpdatesNotWrappedWithActDEV (node_modules/react-dom/cjs/react-dom-client.development.js:15750:9)
      at scheduleUpdateOnFiber (node_modules/react-dom/cjs/react-dom-client.development.js:14311:11)
      at dispatchSetStateInternal (node_modules/react-dom/cjs/react-dom-client.development.js:7208:13)
      at dispatchSetState (node_modules/react-dom/cjs/react-dom-client.development.js:7161:7)
      at /Users/josephine.lee/doula-medicaid-exploration/src/app/form/(formSteps)../../../finish/page.tsx:15:7
```

React `act` points to https://react.dev/reference/react/act (I've never used it before), and mentions that "To avoid some of the boilerplate, you could use a library like [React Testing Library](https://testing-library.com/docs/react-testing-library/intro), whose helpers are wrapped with act()."

This console.error can be solved by deleting this `renderWithRouter()` line, which is okay to delete because we don't need to render the component.

My random hypothesis for why this happens is that React detects that it has rendered a component, then detects that there is an expect without an act (and there is no act because we're not making any RTL assertions).

For the future, I'm not sure how to test that tests don't throw console.errors. We could make a ticket to look into it if you think it's worth doing? (I'm ambivalent either way).
